### PR TITLE
WIP: Add support for Secret Incidents API endpoints

### PR DESCRIPTION
**Pull request by irgeek**
The PR adds models and methods to handle the [Secret Incidents ](https://api.gitguardian.com/docs#tag/Secret-Incidents)API endpoints. It still needs a changelog entry and (probably) some cleanup, but I wanted to get some feedback before I put too much work into it. As requested I've sliced the changes up in to several commits for ease of review.

### CODE REVIEW

**Comment by irgeek** [comment]
The `lint` check is failing because `ggshield` complains about the API key.

File: Pipfile [hunk 1]
```diff
@@ -5,6 +5,7 @@ verify_ssl = true
 
 [packages]
 pygitguardian = { editable = true, path = "." }
+strenum = "*"
```
Comment (agateau-gg) [inline Pipfile:8]: Adding StrEnum here is not required: non-dev dependencies only need to be in `setup.py`.

File: setup.py [hunk 2]
```diff
@@ -34,8 +34,10 @@ def get_version() -> str:
     install_requires=[
         "marshmallow>=3.5, <4",
         "requests>=2, <3",
-        "marshmallow-dataclass >=8.5.8, <8.6.0",
+        "marshmallow-dataclass[enum,union] >=8.5.8, <8.6.0",
         "typing-extensions",
+        "urllib3<2",
+        "strenum",
```
Comment (agateau-gg) [inline setup.py:40]: Can you pin StrEnum to `>=0.4, <0.5` (even if you are very close to StrEnum maintainer 😉)?

File: pygitguardian/incident_models/constants.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+
+class OccurrenceKind(LowercaseStrEnum):
+    REALTIME = auto()
+    HISTORICAL = auto()
+
+
+class OccurrencePresence(SnakeCaseStrEnum):
+    present = auto()
+    removed = auto()
```
Comment (agateau-gg) [inline pygitguardian/incident_models/constants.py:70]: Can you make these values upper-case, like the others?

File: pygitguardian/incident_models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    ignored_at: Optional[datetime]
+    secret_revoked: bool
+    severity: IncidentSeverity = field(metadata={"by_value": True})
+    validity: IncidentValidity = field(metadata={"by_value": True})
+    resolved_at: Optional[datetime]
+    share_url: Optional[str]
+    tags: List[IncidentTag] = field(metadata={"by_value": True})
+
+    def __int__(self):
+        return self.id
```
Comment (agateau-gg) [inline pygitguardian/incident_models/__init__.py:91]: I'd rather not have such magic behavior, using `.id` is simple enough.

File: pygitguardian/utils/tools.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+from typing import Any, Mapping, Sequence
+
+
+def dict_filter_none(dct: Mapping[Any, Any]) -> dict:
+    """Filter a dict to remove all items where the value is None"""
+    return {k: v for k, v in dct.items() if v is not None}
+
+
+def count_not_none(seq: Sequence[Any]) -> int:
+    """Count the number of None values in a sequence"""
```
Comment (agateau-gg) [inline pygitguardian/utils/tools.py:10]:
```suggestion
    """Count the number of not-None values in a sequence"""
```

File: pygitguardian/utils/tools.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    return sum(0 if v is None else 1 for v in seq)
+
+
+def ensure_mutually_exclusive(msg: str, *seq: Any) -> None:
+    """
+    Ensure only one value in a sequence is None.
```
Comment (agateau-gg) [inline pygitguardian/utils/tools.py:16]:
```suggestion
    Ensure only one value in a sequence is not None.
```

File: pygitguardian/utils/tools.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+
+def dict_filter_none(dct: Mapping[Any, Any]) -> dict:
+    """Filter a dict to remove all items where the value is None"""
+    return {k: v for k, v in dct.items() if v is not None}
+
+
+def count_not_none(seq: Sequence[Any]) -> int:
+    """Count the number of None values in a sequence"""
+    return sum(0 if v is None else 1 for v in seq)
```
Comment (agateau-gg) [inline pygitguardian/utils/tools.py:11]: nit: can be slightly simplified:
```suggestion
    return sum(1 for v in seq if v is not None)
```

File: pygitguardian/source_models.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+@dataclass
+class Source(Base):
+    id: int
+    url: str
+    type: str  # TODO: Reserved word
+    full_name: str
+    health: SourceHealth = field(metadata={"by_value": True})
+    open_incidents_count: int  # TODO: Type documented as "number" - what's the difference?
+    closed_incidents_count: int  # TODO: Also "number"
+    visibility: str  # TODO: Really? str
```
Comment (agateau-gg) [inline pygitguardian/source_models.py:45]: Are those TODO actionable? If not, better remove them.

File: pygitguardian/incident_models/__init__.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    detector_group_name: str
+    detector_group_display_name: str
+
+
+DetectorSchema = cast(Type[BaseSchema], class_schema(Detector, BaseSchema))
+Detector.SCHEMA = DetectorSchema()
+
+
+@dataclass
+class Match(Base, FromDictMixin):
```
Comment (agateau-gg) [inline pygitguardian/incident_models/__init__.py:35]: There is already a `Match` class in `models`, can we name this one `OccurrenceMatch`?

**Review by agateau-gg — CHANGES_REQUESTED** [review]
Wow, that's some huge work, thanks!

I made some remarks inline, but here are some more general ones:

I'd rather avoid too deep file hierarchies and make the following changes:
- merge all incident_models code into an incident_models.py module, similar to what you did with source_models.py.
- merge utils/tools.py into utils/__init__.py

I think an utils package should not contain model-specific code so I would move `load_incident_response()` to incident_models.py.

I am not fond of the `_url` param added to `GGClient.request()`. I understand this is required to support paging, but I'd rather not add such a parameter to py-gitguardian public API. Maybe you can take the bulk of the `requests()` method minus the endpoint-to-url code, move it to utils and have both `GGClient.request()` and code using paging to use it?

Some of the functions you added take *a lot* of optional arguments. You should make them keyword-only arguments.

**Comment by agateau-gg** [comment]
Sorry I have to close this PR. It has not moved for several months now. Meanwhile incident support has just been added by #114.

**Pull request closed without merging**